### PR TITLE
feat: LLVM backend: implement the equivalent of -fstack-clash-protection

### PR DESCRIPTION
**Pull request by hargoniX**
Fixes: #2373

### CODE REVIEW

**Comment by bollu** [comment]
I suggest merging this only after #2340 is merged. In theory, this PR should not interfere. In practice, it feels best to not tempt fate.

**Comment by tobiasgrosser** [comment]
#2340 is now complete and pushed (in small pieces) to master. I rebased this PR on master to see if CI still goes through.

File: src/Lean/Compiler/IR/EmitLLVM.lean [hunk 1]
```diff
@@ -27,6 +27,22 @@ namespace LLVM
 -- TODO(bollu): instantiate target triple and find out what size_t is.
 def size_tType (llvmctx : LLVM.Context) : IO (LLVM.LLVMType llvmctx) :=
   LLVM.i64Type llvmctx
+
+-- Helper to add a function if it does not exist, and to return the function handle if it does.
+def getOrAddFunction(m : LLVM.Module ctx) (name : String) (type : LLVM.LLVMType ctx) : BaseIO (LLVM.Value ctx) :=  do
```
Comment (Kha) [inline src/Lean/Compiler/IR/EmitLLVM.lean:32]: Missing space (yes, it was already wrong before)

File: src/Lean/Compiler/IR/EmitLLVM.lean [hunk 2]
```diff
@@ -27,6 +27,22 @@ namespace LLVM
...
+  match (← LLVM.getNamedFunction m name) with
+  | some fn => return fn
+  | none =>
+    let fn ← LLVM.addFunction m name type
+    let attr ← LLVM.createStringAttribute "probe-stack" "inline-asm"
+    LLVM.addAttributeAtIndex fn LLVM.AttributeIndex.AttributeFunctionIndex attr
```
Comment (Kha) [inline src/Lean/Compiler/IR/EmitLLVM.lean:38]: I would add an explanatory comment

**Comment by hargoniX** [comment]
I fixed the review comments and squashed @tobiasgrosser commit in so we can have a bisectable history just in case.

**Comment by tobiasgrosser** [comment]
Thank you @hargoniX

**Pull request merged**